drivers: Optimized periph PWM interfaces [adopted]

gebart opened the pull request:
This is @haukepetersen's #3644, rebased on latest master, squashed and additionally updated the servo API to return void on servo_set in order to match the API change for pwm_set.

There was a merge conflict in the STM32F4 driver stemming from #4454. The change in #4454 was copied with only the change rename the `frequency` variable `freq`

gebart commented:
Reviewing this API change is somewhat low-hanging fruit compared to some of the other periph API changes (e.g. #4430, #3216, #3929, #4040) , since PWM is not widely implemented by the supported CPUs and not widely used either.

haukepetersen commented:
+1 one for the PR, but I can't really ACK it, since most of the changes I did myself...

kYc0o commented:
Can I help to ACK?

gebart commented on drivers/include/periph/pwm.h:
```diff
@@ -18,124 +18,152 @@
- * @return                  -1 on invalid channel
  */
-int pwm_set(pwm_t dev, int channel, unsigned int value);
+void pwm_set(pwm_t dev, uint8_t channel, uint16_t value);
```
Any opinions on limiting `value` to 16 bits?

haukepetersen replied:
yes, I think its good!

A-Paul replied:
What are the advantages?

BytesGalore commented on cpu/lpc11u34/periph/pwm.c:
```diff
@@ -57,15 +57,15 @@ int pwm_init(pwm_t dev, pwm_mode_t mode, unsigned int frequency, unsigned int re
-            PWM_0_DEV->MR1 = (resolution);
-            PWM_0_DEV->MR2 = (resolution);
-            PWM_0_DEV->MR3 = (resolution)-1;
+            PWM_0_DEV->MR0 = (res);
```
why parens here? (and below)

A-Paul commented:
The trailing underscores in the periph_conf.h's guards are addressed in onother PR right?

gebart commented:
@A-Paul There are some PRs open for include guards, but I don't know if these files are fixed there specifically. Regardless, it is better to do a separate PR for the include guards rather than extending the life of this PR..

A-Paul commented:
Has someone already run the test on any HW?

A-Paul commented:
Test on `arduino-due` was successful. STM32F{3,4} -according to @haukepetersen- is O.K. too.

A-Paul commented:
> Can I help to ACK?

@kYc0o, yes sure.

gebart commented:
Rebased, replaced `F_CPU` occurrences by `CLOCK_CORECLOCK`

A-Paul commented on cpu/lpc11u34/periph/pwm.c:
```diff
@@ -40,7 +40,7 @@ int pwm_init(pwm_t dev, pwm_mode_t mode, unsigned int frequency, unsigned int re
             /* Check if the frequency and resolution is applicable */
-            if (CLOCK_CORECLOCK/(resolution*frequency) <= 0) {
+            if (CLOCK_CORECLOCK / (res * freq) <= 0) {
                 return -2;
```
Outdated return value.

A-Paul commented on cpu/lpc11u34/periph/pwm.c:
```diff
@@ -40,7 +40,7 @@ int pwm_init(pwm_t dev, pwm_mode_t mode, unsigned int frequency, unsigned int re
                 return -1;
```
Outdated return value.

A-Paul commented on cpu/lpc11u34/periph/pwm.c:
```diff
@@ -82,7 +82,7 @@ int pwm_init(pwm_t dev, pwm_mode_t mode, unsigned int frequency, unsigned int re
             /* Check if the frequency and resolution is applicable */
-            if (CLOCK_CORECLOCK/(resolution*frequency) <= 0) {
+            if (CLOCK_CORECLOCK / (res * freq) <= 0) {
                 return -2;
```
Outdated return value.

A-Paul commented on cpu/stm32f3/periph/pwm.c:
```diff
@@ -94,12 +97,12 @@ int pwm_init(pwm_t dev, pwm_mode_t mode, unsigned int frequency, unsigned int re
     /* set prescale and auto-reload registers to matching values for resolution and frequency */
-    if ((resolution > 0xffff) || ((resolution * frequency) > pwm_clk)) {
+    if ((res > 0xffff) || ((res * freq) > pwm_clk)) {
         return -2;
```
Outdated return value.

A-Paul commented on cpu/lpc11u34/periph/pwm.c:
```diff
@@ -82,7 +82,7 @@ int pwm_init(pwm_t dev, pwm_mode_t mode, unsigned int frequency, unsigned int re
                 return -1;
```
Outdated return value.

gebart commented:
Rebased. Fixed the return values and added a missing `default: return 0` case for the lpc2387.

OK to squash?

haukepetersen commented:
yes and ACK

haukepetersen commented:
@gebart, have you time to rebase/squash? I could use this PR later today :-)

gebart commented:
@haukepetersen will do right now.

gebart commented:
@haukepetersen I got no conflicts but a 3-way merge in kinetis_common that was automatically resolved by git, I'll re-test with a servo on mulle. Did you test any other platforms?

```
% git rebase master
First, rewinding head to replay your work on top of it...
Applying: tests/periph_pwm: adjusted to interface changes
Applying: drivers/periph/pwm: modernized PWM driver interface
Applying: cpu/lpc11u34: adapted to PWM interface changes
Applying: cpu/lpc2387: adapted to PWM interface changes
Applying: cpu/sam3: adapted to PWM interface changes
Applying: cpu/samd21: adapted to PWM interface changes
Applying: cpu/stm32f3: adapted to PWM interface changes
Applying: cpu/stm32f4: adapted to PWM interface changes
Applying: cpu/kinetis_common: adapted to PWM interface changes
Using index info to reconstruct a base tree...
A	cpu/kinetis_common/pwm.c
Falling back to patching base and 3-way merge...
Auto-merging cpu/kinetis_common/periph/pwm.c
Applying: boards: fixed some PWM definitions
Applying: drivers/servo: adapted to PWM interface changes
Applying: squash cpu/lpc11u34 returns
Applying: squash cpu/lpc2387 returns
Applying: squash cpu/stm32f3 returns
```

gebart commented:
Works on mulle.
@haukepetersen OK to squash?

haukepetersen commented:
nice, thanks, ACK holds

gebart commented:
Travis is green -> GO!
